File: hyemin_3/CircularQueue.java
class CircularQueue extends Queue{	
	int front = 0;
	int rear = 0;
	
	@Override
	void addQ(Value item){
		rear = (rear+1)%QUEUE_SIZE;
		
		if(rear==front){
			System.out.println("queue is full");
			reset();
			return ;
		}
		
		queueItem[rear].value = item.value;
		System.out.println(queueItem[rear].value+"을(를) 추가했습니다.");
		
		System.out.println("front = "+front);
		System.out.println("rear = "+rear);
	}
	
	@Override
	void deleteQ(){
		if(rear==front){
			System.out.println("Queue가 비어있습니다.");
			return ;
		}
		
		else{
			front = (front+1)%QUEUE_SIZE;
			System.out.println(queueItem[front].value+"을(를) 제거했습니다");
			
			System.out.println("front = "+front);
			System.out.println("rear = "+rear);
		}
	}
	
	void reset(){
		if(rear>0) rear--;
		else rear = QUEUE_SIZE-1;
	}
}

File: 2주/circular_queue.cpp
#include <iostream>
using namespace std;
#define N 100

class Q {
public:
	int value;
	int key;
};

Q q[N];

int front = -1;
int rear = -1;

void addq(Q item) {
	rear = (rear+1) % N;
	if (front == rear)
		return;
	q[rear] = item;
	
}

Q deleteq () {
	if (front == rear)
		return;
	front=(front+1) % N;
	return 
		q[front];
}

int main() {
	int n, i;
	Q item[100];
	printf("원소 개수: ");
	scanf("%d", &n);
	printf("값: ");
	for (i=0; i<n; i++) {
		scanf("%d", &item[i].value);
	}

	printf("키: ");
	for (i=0; i<n; i++) {
		scanf("%d", &item[i].key);
	}

	for (i=0; i<n; i++)
		addq(item[i]);

	for (i=0; i<n; i++)
		printf("%d ", deleteq().value);

	printf("\n");
	return 0;
}

